onnxscript/rewriter/pattern_builder_jsm.py

```diff
@@ -0,0 +1,546 @@
+                                             op_type='Gather',
+                                             inputs = [gather_in, gather_index],
+                                             outputs = [gather_out],
+                                             num_outputs = 1
+                                             )
+                                     )
+                squeeze_out.name += "_squeeze_out"
+                squeeze_axis = build_constant_from_tensor(f'{gather_out.name}_squeeze_axis', ir.Tensor(np.array([0])))
+                self.function.append(squeeze_axis)
+                self.function.append(ir.Node(domain='',
+                                             op_type='Squeeze',
+                                             inputs=[gather_out, squeeze_axis.outputs[0]],
+                                             outputs=[squeeze_out],
+                                             num_outputs= 1,
+                                             version = 13
+                                             )
+                                     )
+
+                self.function.sort()
+
+
+    def build_function_match_pattern(self, graph):
+        graph.sort()
+        nodes  = find_nodes_of_optype(graph, self.function.name)
+        nodes.insert(0,graph.node('iteration_ext'))
+        nodes.insert(0,graph.node('condition_ext'))
+
+        ir_model = ir.Model(bGraphView('inlined_pipe_pattern', nodes), ir_version=10)
+
+        model = ir.serde.serialize_model(ir_model)
+        onnx.save(model, 'pipeline_match_pattern.onnx')
+
+        pattern  = direct_convert_ir_graph_to_pattern(ir_model.graph)
+
+        return (pattern, nodes)
+
+
+    def load(self, filename):
+        self._model_proto = onnx.load(filename)
+        self._ir_model    = ir.serde.deserialize_model(self._model_proto)
+        self._ir_graph    = self._ir_model.graph
+
+    def update(self):
+        self._ir_model     = ir.Model(self._ir_graph, ir_version = 10)
+        self._model_proto  = ir.serde.serialize_model(self._ir_model)
+
+    def save(self, filename):
+        self.update()
+        onnx.save(self._model_proto, filename)
+
+    def set_signature_index(self, index, stype):
+        self.signature[index] = stype
+
+    @property
+    def output_signature(self):
+        # The output signature is the same as the input signature but without the iteration input
+        return self.signature[1:]
+
+
+def same(input_list):
+    return len(set(input_list)) == 1
+
+def append_output_to_node(node, output):
+    output._producer = node
+    output._index = node.outputs[-1]._index + 1
+    node._outputs = (*node._outputs, output)
+    node._num_outputs = len(node._outputs)
+
+def prepend_output_to_node(node, output):
+    output._producer = node
+    output._index = 0
+    for outp in node._outputs:
+        outp._index += 1
+    node._outputs = (output, *node._outputs)
+    node._num_outputs = len(node._outputs)
```

Comment:
We should add this to node so there is no need to modify internal states

Reply:
In fact we recommend creating new nodes if you want to update inputs and outputs. This way invariance of the graph is maintained. You may consider accumulating the inputs and outputs into two lists before constructing the node.